openmc/kinetics/solver.py

```diff
@@ -0,0 +1,886 @@
+                ['one_group', 'energy_groups', 'fine_groups', 'tally_groups']):
+            f.require_group(name)
+            f[name].attrs['group_edges'] = groups.group_edges
+            f[name].attrs['num_groups'] = groups.num_groups
+
+        f.attrs['num_delayed_groups'] = self.num_delayed_groups
+        f.attrs['chi_delayed_by_delayed_group'] \
+            = self.chi_delayed_by_delayed_group
+        f.attrs['chi_delayed_by_mesh'] = self.chi_delayed_by_mesh
+        f.attrs['use_agd'] = self.use_agd
+        f.attrs['use_pcmfd'] = self.use_pcmfd
+        f.attrs['num_delayed_groups'] = self.num_delayed_groups
+        f.attrs['num_outer_time_steps'] = self.num_outer_time_steps
+        f.attrs['use_pregenerated_sps'] = self.use_pregenerated_sps
+        f.attrs['constant_seed'] = self.constant_seed
+        f.attrs['seed'] = self.seed
+        f.attrs['core_volume'] = self.core_volume
+        f.attrs['k_crit'] = self.k_crit
+        f.attrs['method'] = self.method
+        f.attrs['pnl'] = self.states['START'].pnl
+        f.require_group('clock')
+        f['clock'].attrs['dt_outer'] = self.clock.dt_outer
+        f['clock'].attrs['dt_inner'] = self.clock.dt_inner
+        f.require_group('OUTER_STEPS')
+        f.require_group('INNER_STEPS')
+        f.create_dataset('beta', data=self.states['START'].beta())
+        f.create_dataset('beta_shape', data=self.states['START'].beta_shape())
+        f.create_dataset('beta_eff', data=self.states['START'].beta_eff)
+        f.create_dataset('power', data=self.states['START'].power)
+        f.create_dataset('decay_rate', data=self.states['START'].decay_rate)
+        f.create_dataset('inverse_velocity', data=self.states['START'].inverse_velocity)
+        total = self.states['START'].absorption + self.states['START'].outscatter
+        f.create_dataset('total', data=total)
+
+        f.close()
+
+    def run_openmc(self, time_point, method='ADIABATIC'):
+
+        self.setup_openmc(time_point, method)
+
+        # Names of the statepoint and summary files
+        sp_old_name = '{}/statepoint.{}.h5'.format(self.directory, self.settings_file.batches)
+        sp_new_name = '{}/statepoint_{:.6f}_sec.{}.h5'\
+                      .format(self.directory, self.clock.times[time_point], self.settings_file.batches)
+        sum_old_name = '{}/summary.h5'.format(self.directory)
+        sum_new_name = '{}/summary_{:.6f}_sec.h5'.format(self.directory, self.clock.times[time_point])
+        old_source = '{}/source.{}.h5'.format(self.directory, self.settings_file.batches)
+        new_source = '{}/source.h5'.format(self.directory)
+
+        # Run OpenMC
+        if not self.use_pregenerated_sps:
+            if self.run_on_cluster:
+
+                # Copy job file to run directory
+                cmd_str = 'cp ' + self.job_file  + ' ' + self.directory
+                subprocess.Popen(cmd_str, shell=True)
+                time.sleep(5)
```

Comment:
We should definitely find a more robust way to ensure this file operation works correctly. I think an alternative may be to open that cp operation process, then have some utility function that checks the file is written. This function could work by running in a loop trying to open the file and exiting once an open succeeds. Of course you would then add a small sleep call in that loop, otherwise it would be nasty to try opening something millions of times in a second.